### DISCUSSION TITLE
What amount of memory do I need for one 512 dimension vector?

### DISCUSSION THREAD

**Discussion by martinenkoEduard** [milvus-io/milvus#18719] [Q&A and General discussion] [answered]
What amount of memory do I need for one 512 dimension vector?

**Comment by martinenkoEduard** [comment]
What is the expected value for  

fields = [
    FieldSchema(name="pk", dtype=DataType.INT64, is_primary=True, auto_id=False),
    FieldSchema(name="embeddings", dtype=DataType.FLOAT_VECTOR, dim=512)
]

for a million entries?

I want to check if I get over usage of memory.

**Comment by martinenkoEduard** [comment]
are there tests?
What are the ways to improve memory performance?

**Reply by martinenkoEduard** [reply]
![Untitled](https://user-images.githubusercontent.com/108512856/185477761-c424eac0-2267-4693-8213-8abef50882a3.jpg)
What does Virt mean in for ETCD? 

It shows huge memory usage, but free -h shows little memory usage...

**Comment by martinenkoEduard** [comment]
Also https://stackoverflow.com/questions/71792879/milvus-has-a-high-virtual-memory-footprint

**Comment by martinenkoEduard** [comment]
In tests I get something around 3.7gb of RAM
for 1 million of 1*512  vectors, on ubuntu 18, deb standalone installation.
But these are random vectors. 

Is it ok or is it too much?

**Comment by martinenkoEduard** [comment]
I haven't tested it with real data yet..

**Comment by martinenkoEduard** [comment]
But it goes up to 5.5 gb. Without reasons.

**Comment by martinenkoEduard** [comment]
Also /usr/bin/milvus uses 1.5 gb of RAM even with a released collection.

**Comment by yhmo** [comment, marked as answer]
Basically, as we know, each dimension is a float(32 bits = 4 Bytes) value. For 512 dimensional vector, each vector requires 512 * 4B = 2KB size. So, 1 million 512 dim vectors totally require 512 * 4B * 1M = 2000MB = 2GB size.

The v2.x has an expending rate for memory usage, we did some test and draw a matrix of the expending rate for each kind index:
![Screenshot from 2022-08-19 14-35-30](https://user-images.githubusercontent.com/2282099/185558308-7c6a5c5c-6a7f-4334-acbd-f44770f46b98.png)

If you didn't create index, it is "FLAT", so the memory usage is about 2Gb * 2 = 4GB.

Some other things that affect memory usage:
- other third-party services(ETCD/Pulsar/MinIO) will occupy extra memory by themself. 
- milvus 2.x is written by Go, sometimes the GC is not ready to free memory after you release a collection.

**Reply by martinenkoEduard** [reply]
Thanks for a great answer!
Does using a different index helps reduce memory consumption? 

So I will be not 2Gb * 2 = 4GB but something 
like 2gb * 1.3 ?

**Reply by yhmo** [reply]
> Thanks for a great answer! Does using a different index helps reduce memory consumption?
> 
> So I will be not 2Gb * 2 = 4GB but something like 2gb * 1.3 ?

You can use the formula to calculate. 
- "nb" is the total count of vectors in the collection
- "d" is the dimension
- "nlist" is a parameter of IVF index
- "m", "nbits" are parameters of PQ index
- "M" is a parameter of HNSW index

**Reply by martinenkoEduard** [reply]
Does data itself stores in memory? Or only index?

Am I right that for flat index 
It will be 
nb * d * 4  * expansion rate = (1000000 * 512 * 4 /1024/1024/1024)   * 2

**Reply by yhmo** [reply]
> Does data itself stores in memory? Or only index?
> 
> Am I right that for flat index It will be nb * d * 4 * expansion rate = (1000000 * 512 * 4 /1024/1024/1024) * 2

Milvus split data into segments, if you didn't call create_index(), the index type is "FLAT", all the segments(we call original data) are in memory. If you called create_index() and all segments have finished building index, then only the index(we call index data) in memory. But if some segments have not finished index, fox example, 100 segments, 40 of them have finished index, then index data of the 40 segments in memory, and original data of the 60 segments in memory. 
If dimension=512, vectors count is 1000000. For "FLAT", required memory size is (1000000 * 512 * 4 /1024/1024/1024) * 2 = 4GB.

**Reply by xiaozhisong** [reply]
ivf_pq  expending rate 40 really ?

**Reply by yhmo** [reply]
> ivf_pq expending rate 40 really ?

Just experience value. Since the theory size of ivf_qp is small, total memory consumption is not large.

**Reply by martinenkoEduard** [reply]
Is FLAT index the smallest index, except for IVF_PQ?

**Comment by martinenkoEduard** [comment]
Additionally have you done any test on hard drive consumption related to the collection size?
This collection with 1 million of entries consumes 37GB of hard disk space...

Most of which are 
/var/lib - 17G
/var/log - 15G

administrator@milvus_2_1_1:/var/lib/milvus$ sudo du -hs * .
970M    etcd-data
13G     minio-data
3.0G    rdb_data
76M     rdb_data_meta_kv

administrator@milvus_2_1_1:/var/log$ sudo du -hs * .
1.9G    journal
13G     syslog

**Reply by yhmo** [reply]
> Additionally have you done any test on hard drive consumption related to the collection size? This collection with 1 million of entries consumes 37GB of hard disk space...
> 
> Most of which are /var/lib - 17G /var/log - 15G
> 
> administrator@milvus_2_1_1:/var/lib/milvus$ sudo du -hs * . 970M etcd-data 13G minio-data 3.0G rdb_data 76M rdb_data_meta_kv
> 
> administrator@milvus_2_1_1:/var/log$ sudo du -hs * . 1.9G journal 13G syslog

Need to clarify something. The `/var/lib` and `/var/log` are not milvus data paths, I think this is not milvus data.
@martinenkoEduard

**Comment by yhmo** [comment]
- Firstly, double-check a configuration in the milvus.yml:
```
dataCoord:
  enableCompaction: true
  enableGarbageCollection: true
```
Make sure the `enableGarbageCollection` is true so that milvus can remove segments(marked as "deleted") permanently. If this item is set to false, and you continually insert vectors, the server will eventually run out of storage.

- Milvus v2.x use MinIO as storage, MinIO service has a replicas number for replication, and the default value is 3? (not sure).

- For cluster, Milvus v2.x use Pulsar as an operation log service, all the insert/delete operations will be stored in Pulsar, which will occupy extra storage space. For standalone, Milvus integrates a rocksdb to do the same with Pulsar, all the insert/delete operations will be stored under rocksdb data path.

- If you created index, index data will occupy more storage space. Different index type has different size.

- Runtime log files also occupy disk space(seems the retention logic is not good, we will optimize it).

For these reasons, milvus v2.x requires more storage size than the original data.
